tokio/src/sync/semaphore_ll.rs

```diff
@@ -449,37 +497,84 @@ impl Semaphore {
         }
     }
 
+    /// The wait node has had all of its permits assigned and has been removed
+    /// from the wait queue.
+    ///
+    /// Attempt to remove the QUEUED bit from the node. If additional permits
+    /// are concurrently requested, the node must be pushed back into the wait
+    /// queued.
+    fn remove_queued(&self, waiter: NonNull<Waiter>, closed: bool) {
+        let mut curr = WaiterState(unsafe { waiter.as_ref() }.state.load(Acquire));
+
+        loop {
+            if curr.is_dropped() {
+                // The Permit dropped, it is on us to release the memory
+                let _ = unsafe { Box::from_raw(waiter.as_ptr()) };
+                return;
+            }
+
+            if curr.permits_to_acquire() > 0 {
```

Comment:
How can it be dequeued then?  And this would then queue a node (which already has some permits) to the end of the queue, right? Maybe it should rather be at the beginning of the queue - or release any previously obtained permits

Reply:
When a node is queued, it cannot be dequeued by the permit handle. It can only be dequeued when `rx_lock` is acquired as part of adding new permits.

Reply:
Might be good if there is a comment on this.